crates/ty_python_semantic/src/ast_node_ref.rs

```diff
@@ -32,54 +32,85 @@ use ruff_db::parsed::ParsedModuleRef;
-    /// the invariant `node belongs to parsed` is upheld.
-    pub(super) unsafe fn new(parsed: ParsedModuleRef, node: &T) -> Self {
+    /// This method may panic or produce unspecified results if the provided module is from a
+    /// different file or Salsa revision than the module to which the node belongs.
+    pub(super) fn new(module_ref: &ParsedModuleRef, node: &T) -> Self {
+        debug_assert_eq!(
+            module_ref.get_by_index(node.node_index()).try_into().ok(),
+            Some(node)
+        );
+
         Self {
-            parsed,
-            node: std::ptr::NonNull::from(node),
+            module_ptr: module_ref.module().as_ptr(),
+            #[cfg(debug_assertions)]
+            kind: AnyNodeRef::from(node).kind(),
+            #[cfg(debug_assertions)]
+            range: node.range(),
+            index: node.node_index().clone(),
+            _node: PhantomData,
         }
     }
 
     /// Returns a reference to the wrapped node.
     ///
-    /// Note that this method will panic if the provided module is from a different file or Salsa revision
-    /// than the module this node was created with.
-    pub fn node<'ast>(&self, parsed: &'ast ParsedModuleRef) -> &'ast T {
-        debug_assert!(Arc::ptr_eq(self.parsed.as_arc(), parsed.as_arc()));
+    /// This method may panic or produce unspecified results if the provided module is from a
+    /// different file or Salsa revision than the module to which the node belongs.
+    pub fn node<'ast>(&self, module_ref: &'ast ParsedModuleRef) -> &'ast T {
+        debug_assert_eq!(module_ref.module().as_ptr(), self.module_ptr);
 
-        // SAFETY: Holding on to `parsed` ensures that the AST to which `node` belongs is still
-        // alive and not moved.
-        unsafe { self.node.as_ref() }
+        // Note that the module pointer is guaranteed to be stable within the Salsa
+        // revision, so the file contents cannot have changed by the above assertion.
+        module_ref
+            .get_by_index(&self.index)
+            .try_into()
+            .ok()
+            .expect("AST indices should never change within the same revision")
     }
 }
 
-impl<T> std::fmt::Debug for AstNodeRef<T>
+#[allow(clippy::missing_fields_in_debug)]
+impl<T> Debug for AstNodeRef<T>
 where
-    T: std::fmt::Debug,
+    T: HasNodeIndex + Debug,
+    for<'ast> &'ast T: TryFrom<AnyRootNodeRef<'ast>>,
 {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_tuple("AstNodeRef")
-            .field(self.node(&self.parsed))
-            .finish()
+        #[cfg(debug_assertions)]
+        {
+            f.debug_struct("AstNodeRef")
+                .field("kind", &self.kind)
+                .field("range", &self.range)
+                .finish()
+        }
+
+        #[cfg(not(debug_assertions))]
+        {
+            // Unfortunately we have no access to the AST here.
+            f.debug_tuple("AstNodeRef").field(&"_").finish()
```

Comment:
Maybe: use `finish_non_exhaustive` to make clear that there's more but we can't show it